ion/src/device/shared/drivers/exam_mode.cpp

```diff
@@ -3,48 +3,38 @@
 #include "flash.h"
 #include <assert.h>
 
-namespace Ion {
-namespace ExamMode {
-
 extern "C" {
   extern char _exam_mode_buffer_start;
   extern char _exam_mode_buffer_end;
 }
 
+namespace Ion {
+namespace ExamMode {
+
 char ones[Config::ExamModeBufferSize]
   __attribute__((section(".exam_mode_buffer")))
   __attribute__((used))
 = {EXAM_BUFFER_CONTENT};
 
 /* The exam mode is written in flash so that it is resilient to resets.
  * We erase the dedicated flash sector (all bits written to 1) and, upon
- * activating or deactivating the exam mode we write one bit to 0. To determine
- * if we are in exam mode, we count the number of leading 0 bits. If it is even,
- * the exam mode is deactivated, if it is odd, the exam mode is activated. */
+ * deactivating or activating standard or Dutch exam mode we write one or two
+ * bits to 0. To determine in which exam mode we are, we count the number of
+ * leading 0 bits. If it is equal to:
+ * - 0[3]: the exam mode is off;
+ * - 1[3]: the standard exam mode is activated;
+ * - 2[3]: the Dutch exam mode is activated. */
 
 /* significantExamModeAddress returns the first uint32_t * in the exam mode
- * flash sector that does not point to 0. If this flash sector has only 0s, it
- * is erased (to 1) and significantExamModeAddress returns the start of the
- * sector. */
+ * flash sector that does not point to 0. If this flash sector has only 0s or
+ * if it has only one 1, it is erased (to 1) and significantExamModeAddress
+ * returns the start of the sector. */
 
-uint32_t * SignificantExamModeAddress() {
-  uint32_t * persitence_start = (uint32_t *)&_exam_mode_buffer_start;
-  uint32_t * persitence_end = (uint32_t *)&_exam_mode_buffer_end;
-  while (persitence_start < persitence_end && *persitence_start == 0x0) {
-    // Skip even number of zero bits
-    persitence_start++;
-  }
-  if (persitence_start == persitence_end) {
-    assert(Ion::Device::Flash::SectorAtAddress((uint32_t)&_exam_mode_buffer_start) >= 0);
-    Ion::Device::Flash::EraseSector(Ion::Device::Flash::SectorAtAddress((uint32_t)&_exam_mode_buffer_start));
-    return (uint32_t *)&_exam_mode_buffer_start;
-  }
-  return persitence_start;
-}
+constexpr static size_t numberOfBitsInByte = 8;
 
-size_t firstOneBit(int i, size_t size) {
+size_t firstOneBitInByte(int i) {
   int minShift = 0;
```

Comment:
Maybe add a comment to showcase what this returns : "if i = 0b00001111111110100101, firstOneBitInByte(i) returns ..."

Reply:
Rename numberOfBitsAfterLeadingZeroes ?